### PR TITLE
Fix job ids

### PR DESCRIPTION
**Pull request by nonword**
Fixes bug where jobId is not generated in any circumstance. Ensures that
jobId is generated when not in a cancellation context. When in a
cancellation context, given jobId is saved with record, but job is not
completed after execution.

Also minor improvements to local dev concerns:
 - lock Node to v10 with .nvmrc
 - add docker-compose for spinning up a seeded pg db
 - add simple event
 - add lib layer
 - note support for `sam` in readme

### CODE REVIEW

**Review by danamansana — APPROVED** [review]

File: src/Controller/CheckinRequestController.php [hunk 1]
```diff
@@ -128,9 +175,11 @@ protected function initiateCheckinRequest(CheckinRequest $checkinRequest)
         $checkinRequest->create();
 
         // Initiate a job for non-cancellation requests.
-        if (is_null($checkinRequest->getJobId()) && $this->isUseJobService()) {
+        if (is_null($checkinRequest->getCancelRequestId()) && $this->isUseJobService()) {
+            // Create a jobId and assign it to special checkinJobId property
+            // (to signal that job should not be completed later in execution)
             $checkinRequest->setCheckinJobId(JobService::generateJobId($this->isUseJobService()));
-            // Set jobId for proper responses for non-cancellation requests.
+            // Copy jobId to jobId property (which is saved in db):
```
Comment (danamansana) [inline src/Controller/CheckinRequestController.php:182]: This differs from `qa`. In `qa`, we have:
```
$checkinRequest->setCheckinJobId($checkinRequest->getJobId());
```
This is to support another service that handles check in requests being able to set the job id. I think we probably want to keep the version in `qa`, but add a line to generate the job id if it doesn't exist, something like:
```
if (is_null($checkinRequest ->getJobId()) {
  $checkinRequest->setJobId(JobService::generateJobId($this->isUseJobService()));
}
$checkinRequest->setCheckinJobId($checkinRequest->getJobId());
```

File: src/Controller/CheckinRequestController.php [hunk 2]
```diff
@@ -165,6 +214,8 @@ protected function updateCheckinRequest(CheckinRequest $checkinRequest, bool $su
         );
 
         // Finish job processing for non-cancellation requests.
+        // Note: Unclear why we don't just check for null
+        // $checkoutRequest->getCancelRequestId() here:
```
Comment (danamansana) [inline src/Controller/CheckinRequestController.php:218]: Following our discussion, I think this is what is going on:

1. We were assuming that `cancelRequestId` being non-null would indicate a cancel request, but being non-null would indicate a checkin request. We are then setting a checkin id

2. We were assuming that a checkin request would come with a `jobId`, which would then be set to the `checkinJobId`. Semantically, we want to finish the job if it is a checkin request, which is why we are checking for a `checkinJobId`.

File: src/Controller/CheckinRequestController.php (hunk 1 above)
Comment (danamansana) [inline src/Controller/CheckinRequestController.php:182]: What I don't understand is why if we had another service handling checkins, we wouldn't ask it to pass a `checkInId`. Then we wouldn't need to assume that the absence of a `cancelId` indicates that this is a checkin, and we should generate a `checkInId`.

**Review by danamansana — CHANGES_REQUESTED** [review]

**Comment by danamansana** [comment]
In summary I think this needs to handle 3 scenarios, of which only the first two were addressed in what exists in `qa`

1. A cancel request. In this case the cancel request id is set, and the checkin id won't be set and the job isn't finished
2. A checkin request handled by another service. In this case the cancel request isn't set, but there is an existing job id. For some reason we don't expect a checkin id. In this case the checkin id is set to the job id, and the job is finished.
3. A checkin request handled solely by this service. In this case, we won't get a job id, so it needs to be set before the checkin id can be set to the job id, but this isn't happening.

**Comment by nonword** [comment]
Seems to me a way to address all three would be something like:

To generate a jobId for scenario 3 (self generated jobId):
```
        if (is_null($checkinRequest->getJobId()) && $this->isUseJobService()) {
            // Create a jobId and assign it to special checkinJobId property
            // (to signal that job should not be completed later in execution)
            $checkinRequest->setCheckinJobId(JobService::generateJobId($this->isUseJobService()));
            // Copy jobId to jobId property (which is saved in db):
            $checkinRequest->setJobId($checkinRequest->getCheckinJobId());
```

To finish the job in scenarios 2 (jobId given as param, but not a cancellation) & 3 (self-generated jobid):
```
        // Finish job processing if:
        //  1. checkinJobId is set (indicating the jobId was generated by us) OR
        //  2. checkinJobId is null (indicating we didn't generate the jobId) and cancelRequestId is null 
        $generatedOwnJobId = !is_null($checkinRequest->getCheckinJobId());
        $jobIdWasGivenButNotACancellation = is_null($checkinRequest->getCheckinJobId()) && is_null($checkinRequest->getCheckinJobId());
        if ($generatedOwnJobId || $jobIdWasGivenButNotACancellation) && $this->isUseJobService()) {
          JobService::finishJob($checkinRequest);
```

That's not pretty. Detecting scenario 2 is the tough one. Incidentally, I think the reason we don't want to identify scenario 2 by a `checkinJobId` parameter (if I understand you right) is that scenario 2 is really one of the checkin-request-sync endpoint being called by the generic proxy-request-consumer, which ideally just re-plays the original async request and all of its parameters unchanged with the sole difference being the added jobId param. Seems tricky to figure out what special param each -sync endpoint needs to identify proxied requests. But maybe a common param like "proxied=true" (which risks colliding with other params).. Or a header..

In the short term we need these endpoints to work for scenarios 1 and 3, so may I suggest I merge these changes as is to QA to clear testing and we address handling scenario 2 later? Scenario 2 will not be needed for these endpoints until we actually wire up proxy-request-consumer to intercept requests on the /checkin-requests endpoint, which isn't going to happen this week

**Comment by danamansana** [comment]
Yes I agree, until we actually have another service using scenario 2, I think it is going to be difficult to decide what should happen in that case

**Review by danamansana — APPROVED** [review]

**Pull request merged**